src/lib/libx6adc.cpp

```diff
@@ -31,191 +41,210 @@ void cleanup(){
 	}
-	return 1;
+	return status;
 }
 
-int initX6(int deviceID) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->init();
+X6_STATUS initX6(int deviceID) {
+	return x6_call(deviceID, &X6_1000::init);
 }
 
-int read_firmware_version(int deviceID) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->read_firmware_version();
+X6_STATUS read_firmware_version(int deviceID, uint32_t* version) {
+	return x6_getter(deviceID, &X6_1000::read_firmware_version, version);
 }
 
-int set_digitizer_mode(int deviceID, int mode) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->set_digitizer_mode(DIGITIZER_MODE(mode));
+X6_STATUS set_digitizer_mode(int deviceID, DigitizerMode mode) {
+	return x6_call(deviceID, &X6_1000::set_digitizer_mode, mode);
 }
 
-int get_digitizer_mode(int deviceID) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return int(X6s_[deviceID]->get_digitizer_mode());
+X6_STATUS get_digitizer_mode(int deviceID, DigitizerMode* digitizerMode) {
+	return x6_getter(deviceID, &X6_1000::get_digitizer_mode, digitizerMode);
 }
 
-int set_sampleRate(int deviceID, double freq){
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->set_clock(X6_1000::INTERNAL_CLOCK, freq); //assume for now we'll use the internal clock
+X6_STATUS set_sampleRate(int deviceID, double freq){
+	//assume for now we'll use the internal clock
+	//varadic pack must be last so pass default arguments here too
+	return x6_call(deviceID, &X6_1000::set_clock, INTERNAL_CLOCK, freq, FRONT_PANEL);
 }
 
-double get_sampleRate(int deviceID) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->get_pll_frequency();
+X6_STATUS get_sampleRate(int deviceID, double* freq) {
+	return x6_getter(deviceID, &X6_1000::get_pll_frequency, freq);
 }
 
-int set_trigger_source(int deviceID, int triggerSource) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->set_trigger_source(X6_1000::TriggerSource(triggerSource));
+X6_STATUS set_trigger_source(int deviceID, TriggerSource triggerSource) {
+	return x6_call(deviceID, &X6_1000::set_trigger_source, triggerSource);
 }
 
-int get_trigger_source(int deviceID) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return int(X6s_[deviceID]->get_trigger_source());
+X6_STATUS get_trigger_source(int deviceID, TriggerSource* triggerSource) {
+	return x6_getter(deviceID, &X6_1000::get_trigger_source, triggerSource);
 }
 
-int set_reference(int deviceID, int referenceSource) {
-	if (!is_open(deviceID)) return X6_1000::DEVICE_NOT_CONNECTED;
-	return X6s_[deviceID]->set_reference(X6_1000::ReferenceSource(referenceSource));
+X6_STATUS set_reference(int deviceID, ReferenceSource src) {
+	//varadic pack must be last so pass default argument here too
+	return x6_call(deviceID, &X6_1000::set_reference, src, 10e6);
```

Comment:
Hmm, that's unfortunate that you have to pass the default value. Otherwise, the templated `x6_call` is nice and clean.